### PR TITLE
[CHK-519] Add retry/circuit breaker (resilience4j)

### PR DESCRIPTION
**Pull request by MarcoAGreco**
#### List of Changes

- Add retry strategy to transaction service
- Add circuit breakers to transaction service
#### Motivation and Context
- Increase availability

The retry logic is the same for all transactionService's methods invoked by the controllers. The circuitBreaker are currently two. One for the operations that require the interaction with the Nodo and the others. The state of the circuit-breakers can be inspected using actuator health endpoint as enabled by the application properties. 

#### How Has This Been Tested?
- Local
To test the implementation I've created a simple http-proxy that intercept the traffic generated by the `transactions-service` that emulate network/application failures (E.g. Empty body with 500 status code).

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/test/java/it/pagopa/transactions/controllers/TransactionsControllerTest.java [hunk 1]
```diff
@@ -378,4 +396,13 @@ void shouldReturnProblemJsonWith400OnBadInput() {
                 .expectBody(ProblemJsonDto.class)
                 .value(p -> assertEquals(400, p.getStatus()));
     }
+
+    @Test
+    void shouldReturnErrorCircuitBreakerOpen(){
+
+        ResponseEntity error = transactionsController.openStateHandler().block();
```
Comment (giovanniberti) [inline src/test/java/it/pagopa/transactions/controllers/TransactionsControllerTest.java:403]: More bikeshedding: what about making this method private and mocking a controller method (any would do) with an exception?

**Review by giovanniberti — DISMISSED** [review]
🚀

**Comment by giovanniberti** [comment]
Is the testing proxy committed anywhere?

**Review by infantesimone — DISMISSED** [review]
ping :-)

File: src/main/java/it/pagopa/transactions/controllers/TransactionsController.java [hunk 2]
```diff
@@ -28,6 +29,12 @@ public class TransactionsController implements TransactionsApi {
     @Autowired
     private TransactionsService transactionsService;
 
+    @ExceptionHandler({ CallNotPermittedException.class })
+    public Mono<ResponseEntity> openStateHandler(){
+        log.error("Error - OPEN circuit breaker");
+        return Mono.just(ResponseEntity.status(503).build());
```
[the remark this answers is not in the document]
Comment (MarcoAGreco) [reply src/main/java/it/pagopa/transactions/controllers/TransactionsController.java:35]: I think that's a good point, I'll update with new body and status code 👍

File: src/test/java/it/pagopa/transactions/controllers/TransactionsControllerTest.java (hunk 1 above)
Comment (MarcoAGreco) [reply src/test/java/it/pagopa/transactions/controllers/TransactionsControllerTest.java:435]: Yes, I'm planning to refactor all tests related to circuit break logic in dedicated PR

Comment (MarcoAGreco) [reply src/test/java/it/pagopa/transactions/controllers/TransactionsControllerTest.java:435]: > Is the testing proxy committed anywhere?

Nope, it's a node app develop with http-proxy lib

File: pom.xml [hunk 3]
```diff
@@ -153,7 +153,24 @@
...
+            <groupId>io.github.resilience4j</groupId>
+            <artifactId>resilience4j-spring-boot2</artifactId>
+            <version>1.7.1</version>
+        </dependency>
+        <dependency>
+            <groupId>io.github.resilience4j</groupId>
+            <artifactId>resilience4j-reactor</artifactId>
+            <version>1.7.1</version>
+        </dependency>
+        <dependency>
```
Comment (infantesimone) [inline pom.xml:166]: duplicate :-) @MarcoAGreco

**Review by infantesimone — DISMISSED** [review]
just a deps duplicate

File: pom.xml (hunk 3 above)
Comment (MarcoAGreco) [reply pom.xml:166]: Thank you :) solved with [1bcd170](https://github.com/pagopa/pagopa-ecommerce-transactions-service/pull/79/commits/1bcd170fedf1812711403b37aaea80c8be58f3d1)

**Review by infantesimone — APPROVED** [review]

**Pull request merged**